### PR TITLE
avm2: implemented flash.utils.unescapeMultiByte

### PR DESCRIPTION
**Pull request by dowgird**

### CODE REVIEW

File: core/src/avm2/globals/flash/utils.rs [hunk 1]
```diff
@@ -140,6 +140,56 @@ pub fn escape_multi_byte<'gc>(
...
+) -> Result<Value<'gc>, Error> {
+    let s = args
+        .get(0)
+        .unwrap_or(&Value::Undefined)
+        .coerce_to_string(activation)?;
+    let bs = s.as_wstr();
+    let mut buf = WString::new();
+    let mut it = bs.chars().map(|c| c.unwrap_or(char::REPLACEMENT_CHARACTER));
+
+    let mut bytes = Vec::new();
```
Comment (EmperorBale) [inline core/src/avm2/globals/flash/utils.rs:166]: Instead of allocating a new buffer everytime we encounter a `%`, is there a way we can write the bytes directly into the WString without this?

Comment (dowgird) [reply core/src/avm2/globals/flash/utils.rs:166]: The main problem is that the input string can mix percent-escaped bytes with unescaped chars, including wide ones. So what we write to the string must be already normalized to wide chars.

These runs of bytes need to be converted to utf-16. It is possible to avoid the allocation and have a fixed buffer (like 4 bytes), then accumulate the bytes until either:

* they form a valid utf-8 encoding of a char
* they form an invalid utf-8 sequence
* an unescaped char comes next
* the input string ends

Basically create a version of`DecodeAvmUtf8` which accepts bytes one by one.

My opinion: only makes sense if this mechanism (i mean "decode utf-8 as you go") is usable elsewhere, otherwise probably not worth it.

File: core/src/avm2/globals/flash/utils.rs [hunk 2]
```diff
@@ -140,6 +140,56 @@ pub fn escape_multi_byte<'gc>(
     Ok(v.into())
 }
 
+fn handle_percent<I>(it: &mut I) -> Option<u8>
+where
+    I: Iterator<Item = char>,
+{
+    let mut d = it.next()?.to_digit(16)?;
+    d = d * 16 + it.next()?.to_digit(16)?;
```
Comment (relrelb) [inline core/src/avm2/globals/flash/utils.rs:148]: nit: This can be cleaned up a bit (no `mut`, bitwise instead of arithmetic operations):
```rs
let high = it.next()?.to_digit(16)? as u8;
let low = it.next()?.to_digit(16)? as u8;
Some(low | (high << 4))
```

File: core/src/avm2/globals/flash/utils.rs [hunk 3]
```diff
@@ -140,6 +140,56 @@ pub fn escape_multi_byte<'gc>(
     Ok(v.into())
 }
 
+fn handle_percent<I>(it: &mut I) -> Option<u8>
```
Comment (relrelb) [inline core/src/avm2/globals/flash/utils.rs:143]: `handle_percent` can be defined as a closure inside `unescape_multi_byte`, so that the `it` parameter and generic won't be needed.

File: core/src/avm2/globals/flash/utils.rs [hunk 4]
```diff
@@ -140,6 +140,56 @@ pub fn escape_multi_byte<'gc>(
...
+    _this: Option<Object<'gc>>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    let s = args
+        .get(0)
+        .unwrap_or(&Value::Undefined)
+        .coerce_to_string(activation)?;
+    let bs = s.as_wstr();
+    let mut buf = WString::new();
+    let mut it = bs.chars().map(|c| c.unwrap_or(char::REPLACEMENT_CHARACTER));
```
Comment (relrelb) [inline core/src/avm2/globals/flash/utils.rs:164]: nit: Let's rename `it` to `chars`, to be consistent with `effective_replacement` in `regexp.rs`.

File: core/src/avm2/globals/flash/utils.rs (hunk 1 above)
Comment (relrelb) [reply core/src/avm2/globals/flash/utils.rs:166]: I think it's good enough to keep as-is. But for code clarity, I think `bytes` can be defined in a narrower scope, if we'd have peekable chars iterator. Something like:
```rs
// ...
if c == '%' {
    let mut bytes = Vec::new();
    while let Some(b) = handle_percent() {
        bytes.push(b);
        if !matches!(chars.peek(), Some('%')) {
            break;
        }
    }
    buf.push_str(&WString::from_utf8_bytes(bytes));
}
// ...
```

File: core/src/avm2/globals/flash/utils.rs (hunk 4 above)
Comment (dowgird) [reply core/src/avm2/globals/flash/utils.rs]: Done.

File: core/src/avm2/globals/flash/utils.rs (hunk 2 above)
Comment (dowgird) [reply core/src/avm2/globals/flash/utils.rs]: Done.

File: core/src/avm2/globals/flash/utils.rs (hunk 1 above)
Comment (dowgird) [reply core/src/avm2/globals/flash/utils.rs]: Done.

File: core/src/avm2/globals/flash/utils.rs (hunk 3 above)
Comment (dowgird) [reply core/src/avm2/globals/flash/utils.rs]: The closure borrows the iterator mutably. This clashes with the main loop also using it mutably.

**Review by relrelb — APPROVED** [review]
Thank you!
